### PR TITLE
Implemented first version of load test

### PR DESCRIPTION
**Pull request by piosz**
As for now it creates a bunch of Replication Controllers and play with them

Partially addresses #3139

### CODE REVIEW

**Comment by piosz** [comment]
cc @davidopp @saad-ali @quinton-hoole @jayunit100

**Comment by quinton-hoole** [comment]
cc @bprashanth who knows where many of the bodies are buried w.r.t. replication controllers, and might want to take a look over this code too, and comment on whether it covers the important cases.  I'll also review the code shortly.

File: test/e2e/load.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+var _ = Describe("Load", func() {
+	var c *client.Client
+	var minionCount int
+	var ns string
+
+	BeforeEach(func() {
+		var err error
+		c, err = loadClient()
+		expectNoError(err)
+		minions, err := c.Nodes().List(labels.Everything(), fields.Everything())
```
Comment (fgrzadkowski) [inline test/e2e/load.go]: s/minions/nodes

File: test/e2e/load.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+		expectNoError(err)
+		minionCount = len(minions.Items)
```
Comment (fgrzadkowski) [inline test/e2e/load.go]: nodesCount

File: test/e2e/load.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+			runRCLoad(c, &wg, ns, bigRCSize, bigRCCount)
+
+			// Wait for all the pods from all the RC's to return.
+			wg.Wait()
+		})
+	}
+})
+
+func computeRCCounts(total int) (int, int, int) {
+	// Small RCs owns ~0.5 of total number of pods, medium and big RCs ~0.25 each.
```
Comment (fgrzadkowski) [inline test/e2e/load.go]: Can you please expand this comment to show some concrete numbers? e.g. For 1000 pods we will have:
* X small RC each 5 pods
* Y medium RC each ...

File: test/e2e/load.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+	bigRCCount := total / 4 / bigRCSize
```
Comment (fgrzadkowski) [inline test/e2e/load.go]: I think it will be more readable to write:

smallCount := total * bigShare / bigSize
mediumCount := total * mediumShare / mediumSize
...

File: test/e2e/load.go [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+	mediumRCCount := (total - bigRCCount*bigRCSize) / 3 / mediumRCSize
+	smallRCCount := (total - bigRCCount*bigRCSize - mediumRCCount*mediumRCSize) / smallRCSize
+	return smallRCCount, mediumRCCount, bigRCCount
+}
+
+// The function creates a RC and then every few second resize it and with 0.1 probability deletes it.
+func playWithRC(c *client.Client, wg *sync.WaitGroup, ns string, size int) {
```
Comment (fgrzadkowski) [inline test/e2e/load.go]: How can we debug this test if this is random?

File: test/e2e/load.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+	defer GinkgoRecover()
```
Comment (fgrzadkowski) [inline test/e2e/load.go]: Why is this needed?

File: test/e2e/load.go [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+	smallRCSize    = 5
+	mediumRCSize   = 30
+	bigRCSize      = 250
+)
+
+// This test suite can take a long time to run, so by default it is added to
+// the ginkgo.skip list (see driver.go).
+// To run this suite you must explicitly ask for it by setting the
+// -t/--test flag or ginkgo.focus flag.
+var _ = Describe("Load", func() {
```
Comment (fgrzadkowski) [inline test/e2e/load.go]: Please add a TODO somewhere to verify latency metrics

File: test/e2e/load.go [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+	defer wg.Done()
+	rcExist := false
+	var name string
+	for start := time.Now(); time.Since(start) < simulationTime; time.Sleep(time.Duration(rand.Intn(10)) * time.Second) {
+		if !rcExist {
+			name = "load-test-" + string(util.NewUUID())
+			expectNoError(RunRC(c, name, ns, image, size))
+			rcExist = true
+		}
+		// Resize RC to a random size between 0.5x and 1.5x of the original size.
```
Comment (fgrzadkowski) [inline test/e2e/load.go]: One concern I have is that we will generate too much load with this test. How many requests does it generate? I think that in reality we will not see resizing every few seconds. Maybe let's resize each RC only few times during the test (5 min)?

File: test/e2e/load.go [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+const (
+	image          = "gcr.io/google_containers/serve_hostname:1.1"
+	simulationTime = 5 * time.Minute
```
Comment (fgrzadkowski) [inline test/e2e/load.go]: Is this enough? In density tests starting 3000 pods takes ~9.5 min until they are running. Here you also want to run 3000 pods and then resize multiple for some time. Maybe it'd be better to start the time only when all pods are running. WDYT?

File: test/e2e/load.go [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+	bigRCSize      = 250
+)
+
+// This test suite can take a long time to run, so by default it is added to
+// the ginkgo.skip list (see driver.go).
+// To run this suite you must explicitly ask for it by setting the
+// -t/--test flag or ginkgo.focus flag.
+var _ = Describe("Load", func() {
+	var c *client.Client
+	var minionCount int
```
Comment (davidopp) [inline test/e2e/load.go]: s/minionCount/nodeCount/

File: test/e2e/load.go (hunk 5 above)
Comment (davidopp) [inline test/e2e/load.go]: My suggestion was that the first version of the load generator could be random, and then later we could enhance it to also test correctness. This isn't ideal, but given our time constraints and the fact that it's a lot easier to do random, I think it makes sense to do random first.

**Comment by davidopp** [comment]
It would be great to have a mode for this test where it just prints out what it would do, rather than actually doing it (like "dry run" mode for various Borg tools). This would make it easier for people to understand how the test is expected to behave, which isn't completely obvious just by looking at the code, due to the random numbers, scaling factors, etc.

File: test/e2e/load.go [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+		minionCount = len(minions.Items)
+		Expect(minionCount).NotTo(BeZero())
+		nsForTesting, err := createTestingNS("load", c)
+		ns = nsForTesting.Name
+		expectNoError(err)
+	})
+
+	AfterEach(func() {
+		By(fmt.Sprintf("Destroying namespace for this suite %v", ns))
+		if err := c.Namespaces().Delete(ns); err != nil {
```
Comment (bprashanth) [inline test/e2e/load.go]: how hard is it to pipe through a flag that skips cleanup on failure? i often leave e2e stress tests overnight and would find that very useful.

File: test/e2e/load.go [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+		It(name, func() {
+			totalPods := testArg.podsPerMinion * minionCount
+			smallRCCount, mediumRCCount, bigRCCount := computeRCCounts(totalPods)
+			threads := smallRCCount + mediumRCCount + bigRCCount
+
+			var wg sync.WaitGroup
+			wg.Add(threads)
+
+			// Run RC load for all kinds of RC.
+			runRCLoad(c, &wg, ns, smallRCSize, smallRCCount)
```
Comment (bprashanth) [inline test/e2e/load.go]: If one of these fails, I'd really like to know what the state of it's pods are (i.e how many, on what minions, in what state). Thoughts?

File: test/e2e/util.go [hunk 13]
```diff
@@ -571,6 +536,68 @@ func RunRC(c *client.Client, name string, ns, image string, replicas int) error
 	return nil
 }
 
+func ResizeRC(c *client.Client, ns, name string, size int) error {
+	By(fmt.Sprintf("Resizing replication controller %s to %d", name, size))
```
Comment (bprashanth) [inline test/e2e/util.go]: If you could reuse the kubectl libraries: 
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubectl/resize.go
that would be ideal

eg: 
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubectl/stop.go#L82

File: test/e2e/util.go [hunk 14]
```diff
@@ -571,6 +536,68 @@ func RunRC(c *client.Client, name string, ns, image string, replicas int) error
...
+		err := wait.Poll(time.Second, remainingTime, client.ControllerHasDesiredReplicas(c, rc))
+		if err != nil {
+			Logf("Error while waiting for replication controller %s replicas to read %v: %s", name, size, err)
+		} else {
+			break
+		}
+	}
+
+	// Wait up to 10 minutes for pods to become Running.
+	running := false
```
Comment (bprashanth) [inline test/e2e/util.go]: suggest splitting this out into another function that just waits for pods so we can, say, wait for 50% of the pods or something down the line.

File: test/e2e/util.go [hunk 15]
```diff
@@ -571,6 +536,68 @@ func RunRC(c *client.Client, name string, ns, image string, replicas int) error
...
+	label := labels.SelectorFromSet(labels.Set(map[string]string{"name": name}))
+	for start := time.Now(); time.Since(start) < 10*time.Minute; time.Sleep(5 * time.Second) {
```
Comment (bprashanth) [inline test/e2e/util.go]: Listing all pods every 5s was exactly the thing a lot of our work was done to avoid, since it places quite a bit of load on the system by itself :)

Unsure how much context you have about similar code, but another way to achieve this would be to setup a reflector, eg: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubelet/config/apiserver.go#L43. A controller is an even higher abstraction, but I think a reflector should do for this. You can then poll the store.

File: test/e2e/util.go [hunk 16]
```diff
@@ -571,6 +536,68 @@ func RunRC(c *client.Client, name string, ns, image string, replicas int) error
...
+	}
+	if !running {
+		return fmt.Errorf("Timeout while waiting for replication controller %s pods to be running", name)
+	}
+	return nil
+}
+
+// Delete a Replication Controller and all pods it spawned
+func DeleteRC(c *client.Client, ns, name string) error {
+	// Resize the replication controller to 0.
```
Comment (bprashanth) [inline test/e2e/util.go]: you should be able to use the kubectl stop lib mentioned in my previous comment

**Comment by bprashanth** [comment]
@piosz thanks! I think this will be helpful

File: test/e2e/load.go (hunk 6 above)
Comment (piosz) [inline test/e2e/load.go]: https://github.com/onsi/ginkgo/blob/master/ginkgo_dsl.go#L37

File: test/e2e/load.go (hunk 9 above)
Comment (piosz) [inline test/e2e/load.go]: I'll change to 20 minutes by default.

File: test/e2e/load.go (hunk 7 above)
Comment (piosz) [inline test/e2e/load.go]: Done

File: test/e2e/load.go (hunk 10 above)
Comment (piosz) [inline test/e2e/load.go]: Done

File: test/e2e/load.go (hunk 1 above)
Comment (piosz) [inline test/e2e/load.go]: Done

File: test/e2e/load.go (hunk 2 above)
Comment (piosz) [inline test/e2e/load.go]: Done

File: test/e2e/load.go (hunk 11 above)
Comment (piosz) [inline test/e2e/load.go]: I don't know. I'll add TODO. It may be worth to add such flag also for other stress tests.

File: test/e2e/load.go (hunk 5 above)
Comment (piosz) [inline test/e2e/load.go]: Agree with @davidopp. I'll add verification and make the test deterministic later.

File: test/e2e/load.go (hunk 8 above)
Comment (piosz) [inline test/e2e/load.go]: You are right. I'll change resizing to happen once every 1-2 minutes.

File: test/e2e/load.go (hunk 3 above)
Comment (piosz) [inline test/e2e/load.go]: Done

File: test/e2e/load.go (hunk 4 above)
Comment (piosz) [inline test/e2e/load.go]: I agree that current implementation is a little bit confusing, but adding share factor will require converting the numbers between float and int representation.

Also the current implementation guarantees that the total number of pods will be (almost) equal to the requested "total".

File: test/e2e/util.go (hunk 13 above)
Comment (piosz) [inline test/e2e/util.go]: Done

File: test/e2e/util.go (hunk 16 above)
Comment (piosz) [inline test/e2e/util.go]: Done

File: test/e2e/util.go (hunk 15 above)
Comment (piosz) [inline test/e2e/util.go]: There is the same pattern in RunRC function. Since nobody complains let's leave it as it is now. When we decide to get rid of listing pods from e2e tests we will do it in both places.

File: test/e2e/util.go (hunk 14 above)
Comment (piosz) [inline test/e2e/util.go]: Done

**Comment by piosz** [comment]
Addressed comments and rebased. PTAL

I'll squash before merge

File: test/e2e/util.go [hunk 17]
```diff
@@ -576,6 +540,53 @@ func RunRC(c *client.Client, name string, ns, image string, replicas int) error
...
+		break
+	}
+	if !running {
+		return fmt.Errorf("Timeout while waiting for replication controller %s pods to be running", rcName)
+	}
+	return nil
+}
+
+// Delete a Replication Controller and all pods it spawned
+func DeleteRC(c *client.Client, ns, name string) error {
```
Comment (bprashanth) [inline test/e2e/util.go]: To avoid confusing why not just call this stop? since there is a delete operation as well, but it doesn't resize. can you also log the delete rc name/namespace since you do it randomly above?

File: test/e2e/util.go [hunk 18]
```diff
@@ -576,6 +540,53 @@ func RunRC(c *client.Client, name string, ns, image string, replicas int) error
 	return nil
 }
 
+func ResizeRC(c *client.Client, ns, name string, size uint) error {
+	By(fmt.Sprintf("Resizing replication controller %s to %d", name, size))
+	resizer, err := kubectl.ResizerFor("ReplicationController", kubectl.NewResizerClient(c))
+	if err != nil {
+		return err
+	}
+	waitForReplicas := kubectl.NewRetryParams(5*time.Second, 5*time.Minute)
```
Comment (bprashanth) [inline test/e2e/util.go]: Either expose this so I can control the resize to return without waiting, or force everyone to wait and delete waitForRCPodsRunning? Otherwise I'm not sure how I'd use the 2 together

File: test/e2e/util.go [hunk 19]
```diff
@@ -576,6 +540,53 @@ func RunRC(c *client.Client, name string, ns, image string, replicas int) error
 	return nil
 }
 
+func ResizeRC(c *client.Client, ns, name string, size uint) error {
+	By(fmt.Sprintf("Resizing replication controller %s to %d", name, size))
```
Comment (bprashanth) [inline test/e2e/util.go]: please also include the ns

File: test/e2e/util.go (hunk 17 above)
Comment (piosz) [inline test/e2e/util.go]: AFAIK we want to get rid of stop, so let's call this delete.

For more context https://github.com/GoogleCloudPlatform/kubernetes/pull/7432#discussion_r29259589

File: test/e2e/util.go (hunk 19 above)
Comment (piosz) [inline test/e2e/util.go]: Done

File: test/e2e/util.go (hunk 18 above)
Comment (piosz) [inline test/e2e/util.go]: I think we should keep such methods as simple as possible ie. when someone is writing e2e test he shouldn't care about waiting timeout. Also it's only e2e so we don't care that much about performance, so let's leave it as it is now. Function waitForRCPodsRunning is just factored out from ResizeRC to make code more readable and may be used in a different methods in util.go.

File: test/e2e/util.go [hunk 20]
```diff
@@ -576,6 +540,54 @@ func RunRC(c *client.Client, name string, ns, image string, replicas int) error
...
+	}
+	waitForReplicas := kubectl.NewRetryParams(5*time.Second, 5*time.Minute)
+	if err = resizer.Resize(ns, name, size, nil, nil, waitForReplicas); err != nil {
+		return err
+	}
+	return waitForRCPodsRunning(c, ns, name)
+}
+
+func waitForRCPodsRunning(c *client.Client, ns, rcName string) error {
+	// Wait up to 10 minutes for pods to become Running.
```
Comment (fgrzadkowski) [inline test/e2e/util.go]: Please move this comment before function header.

File: test/e2e/util.go [hunk 21]
```diff
@@ -576,6 +540,54 @@ func RunRC(c *client.Client, name string, ns, image string, replicas int) error
...
+	By(fmt.Sprintf("Resizing replication controller %s in namespace %s to %d", name, ns, size))
+	resizer, err := kubectl.ResizerFor("ReplicationController", kubectl.NewResizerClient(c))
+	if err != nil {
+		return err
+	}
+	waitForReplicas := kubectl.NewRetryParams(5*time.Second, 5*time.Minute)
+	if err = resizer.Resize(ns, name, size, nil, nil, waitForReplicas); err != nil {
+		return err
+	}
+	return waitForRCPodsRunning(c, ns, name)
```
Comment (fgrzadkowski) [inline test/e2e/util.go]: Can we be sure that before we call waitForRCPdosRunning all pods are already created/deleted?

File: pkg/kubectl/resize.go [hunk 22]
```diff
@@ -105,6 +105,10 @@ type RetryParams struct {
 	interval, timeout time.Duration
 }
 
+func NewRetryParams(interval, timeout time.Duration) *RetryParams {
```
Comment (fgrzadkowski) [inline pkg/kubectl/resize.go]: Is it needed? Calling this function is similarly long to just creating struct.

Comment (piosz) [inline pkg/kubectl/resize.go]: As we discussed it seems to be.

File: test/e2e/util.go (hunk 20 above)
Comment (piosz) [inline test/e2e/util.go]: Done

File: test/e2e/util.go (hunk 21 above)
Comment (piosz) [inline test/e2e/util.go]: Yes, otherwise resizer.Resize returns error.

**Comment by piosz** [comment]
Rebased and squashed.

**Comment by fgrzadkowski** [comment]
LGTM.
